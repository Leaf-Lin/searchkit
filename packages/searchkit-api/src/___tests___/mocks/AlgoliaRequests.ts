export const SimpleRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facetFilters: [['type:movie']],
      facets: ['*'],
      highlightPostTag: '</ais-highlight-0000000000>',
      highlightPreTag: '<ais-highlight-0000000000>',
      maxValuesPerFacet: 10,
      page: 0,
      query: 'shawshank',
      tagFilters: ''
    }
  }
]

export const nonDynamicFacetRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facets: ['type', 'actors', 'imdbrating'],
      highlightPostTag: '</ais-highlight-0000000000>',
      highlightPreTag: '<ais-highlight-0000000000>',
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const nonDynamicFacetRequestOneFilter = [
  {
    indexName: 'imdb_movies',
    params: {
      facetFilters: [['actors:Alice']],
      facets: ['*'],
      highlightPostTag: '</ais-highlight-0000000000>',
      highlightPreTag: '<ais-highlight-0000000000>',
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const DisjunctiveExampleRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facetFilters: [['type:movie']],
      facets: ['*'],
      highlightPostTag: '</ais-highlight-0000000000>',
      highlightPreTag: '<ais-highlight-0000000000>',
      maxValuesPerFacet: 10,
      page: 0,
      query: 'shawshank',
      tagFilters: ''
    }
  },
  {
    indexName: 'imdb_movies',
    params: {
      analytics: false,
      attributesToHighlight: [],
      attributesToRetrieve: [],
      attributesToSnippet: [],
      clickAnalytics: false,
      facets: 'type',
      highlightPostTag: '</ais-highlight-0000000000>',
      highlightPreTag: '<ais-highlight-0000000000>',
      hitsPerPage: 1,
      maxValuesPerFacet: 10,
      page: 0,
      query: 'shawshank',
      tagFilters: ''
    }
  }
]

export const NestedQueryRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facetFilters: [['user.first:Alice']],
      facets: ['*'],
      highlightPostTag: '</ais-highlight-0000000000>',
      highlightPreTag: '<ais-highlight-0000000000>',
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  },
  {
    indexName: 'imdb_movies',
    params: {
      analytics: false,
      attributesToHighlight: [],
      attributesToRetrieve: [],
      attributesToSnippet: [],
      clickAnalytics: false,
      facets: 'user.first',
      highlightPostTag: '</ais-highlight-0000000000>',
      highlightPreTag: '<ais-highlight-0000000000>',
      hitsPerPage: 1,
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const TwoFiltersNestedQueryRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facetFilters: [['user.first:Alice', 'user.first:John']],
      facets: ['*'],
      highlightPostTag: '</ais-highlight-0000000000>',
      highlightPreTag: '<ais-highlight-0000000000>',
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  },
  {
    indexName: 'imdb_movies',
    params: {
      analytics: false,
      attributesToHighlight: [],
      attributesToRetrieve: [],
      attributesToSnippet: [],
      clickAnalytics: false,
      facets: 'user.first',
      highlightPostTag: '</ais-highlight-0000000000>',
      highlightPreTag: '<ais-highlight-0000000000>',
      hitsPerPage: 1,
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const NumericFilterNestedQueryRequest = [
  {
    indexName: 'my-index-000001',
    params: {
      facetFilters: [],
      facets: ['*'],
      highlightPostTag: '</ais-highlight-0000000000>',
      highlightPreTag: '<ais-highlight-0000000000>',
      maxValuesPerFacet: 10,
      numericFilters: ['user.price>=90'],
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const FacetQueryExampleRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facetName: 'type',
      facetQuery: 'm',
      facets: ['*'],
      highlightPostTag: '</ais-highlight-0000000000>',
      highlightPreTag: '<ais-highlight-0000000000>',
      maxFacetHits: 10,
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const NumericFiltersExampleRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facets: ['type', 'actors', 'imdbrating'],
      highlightPostTag: '</ais-highlight-0000000000>',
      highlightPreTag: '<ais-highlight-0000000000>',
      maxValuesPerFacet: 10,
      numericFilters: ['imdbrating>=90'],
      page: 0,
      query: '',
      tagFilters: ''
    }
  },
  {
    indexName: 'imdb_movies',
    params: {
      analytics: false,
      clickAnalytics: false,
      facets: 'imdbrating',
      highlightPostTag: '</ais-highlight-0000000000>',
      highlightPreTag: '<ais-highlight-0000000000>',
      hitsPerPage: 0,
      maxValuesPerFacet: 10,
      page: 0,
      query: ''
    }
  }
]
