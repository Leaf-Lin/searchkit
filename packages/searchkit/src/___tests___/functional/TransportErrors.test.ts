import Client from '../../'
import { DisjunctiveExampleRequest } from '../mocks/AlgoliaRequests'
import type { AlgoliaMultipleQueriesQuery } from '@searchkit/api'

import nock from 'nock'
import { HitsResponseWithFacetFilter } from '../mocks/ElasticsearchResponses'
import {
  APIKeyNotAuthorized,
  ElasticsearchError,
  InvalidRequest,
  KeyUnauthorizedForIndex
} from '../mocks/ErrorElasticsearchResponses'

describe('Transport Errors', () => {
  const client = new Client({
    connection: {
      host: 'https://commerce-demo.es.us-east4.gcp.elastic-cloud.com:9243',
      apiKey: 'apikey'
    },
    search_settings: {
      highlight_attributes: ['title', 'actors'],
      search_attributes: ['title', 'actors', 'query'],
      result_attributes: ['title', 'actors', 'query'],
      facet_attributes: ['type'],
      query_rules: []
    }
  })

  it('401 error', async () => {
    nock('https://commerce-demo.es.us-east4.gcp.elastic-cloud.com:9243')
      .post('/_msearch', (requestBody: any) => {
        return true
      })
      .reply(200, APIKeyNotAuthorized)

    await expect(() => {
      return client.handleInstantSearchRequests(
        DisjunctiveExampleRequest as AlgoliaMultipleQueriesQuery[]
      )
    }).rejects.toThrowErrorMatchingSnapshot()
  })

  it('403 error', async () => {
    nock('https://commerce-demo.es.us-east4.gcp.elastic-cloud.com:9243')
      .post('/_msearch', (requestBody: any) => {
        return true
      })
      .reply(200, KeyUnauthorizedForIndex)

    await expect(() => {
      return client.handleInstantSearchRequests(
        DisjunctiveExampleRequest as AlgoliaMultipleQueriesQuery[]
      )
    }).rejects.toThrowErrorMatchingSnapshot()
  })

  it('400 error', async () => {
    nock('https://commerce-demo.es.us-east4.gcp.elastic-cloud.com:9243')
      .post('/_msearch', (requestBody: any) => {
        return true
      })
      .reply(200, InvalidRequest)

    await expect(() => {
      return client.handleInstantSearchRequests(
        DisjunctiveExampleRequest as AlgoliaMultipleQueriesQuery[]
      )
    }).rejects.toThrowErrorMatchingSnapshot()
  })

  it('500 error', async () => {
    nock('https://commerce-demo.es.us-east4.gcp.elastic-cloud.com:9243')
      .post('/_msearch', (requestBody: any) => {
        return true
      })
      .reply(200, ElasticsearchError)

    await expect(() => {
      return client.handleInstantSearchRequests(
        DisjunctiveExampleRequest as AlgoliaMultipleQueriesQuery[]
      )
    }).rejects.toThrowErrorMatchingSnapshot()
  })

  it('200 but bad response', async () => {
    nock('https://commerce-demo.es.us-east4.gcp.elastic-cloud.com:9243')
      .post('/_msearch', (requestBody: any) => {
        return true
      })
      .reply(200, {
        responses: [{ hits: [] }],
        status: 200
      })

    await expect(() => {
      return client.handleInstantSearchRequests(
        DisjunctiveExampleRequest as AlgoliaMultipleQueriesQuery[]
      )
    }).rejects.toThrowErrorMatchingSnapshot()
  })
})
