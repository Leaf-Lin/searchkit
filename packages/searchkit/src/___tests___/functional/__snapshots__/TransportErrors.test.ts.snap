// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transport Errors 200 but bad response 1`] = `"Error transforming Elasticsearch response for index"`;

exports[`Transport Errors 400 error 1`] = `"Elasticsearch Bad Request. Check your query and make sure it is valid. Check the field mapping. Turn on debug mode to see the Elasticsearch query."`;

exports[`Transport Errors 401 error 1`] = `"Cannot connect to Elasticsearch. Check your connection host and API Key. You can also provide a custom Elasticsearch transporter to the API Client. See docs for more information."`;

exports[`Transport Errors 403 error 1`] = `"Auth Error: You do not have permission to access this index. Check you are calling the right index (specified in frontend) and your API Key permissions has access to the index."`;

exports[`Transport Errors 500 error 1`] = `"Elasticsearch Internal Error: Check your elasticsearch instance to make sure it can recieve requests."`;
